### ISSUE TITLE
[Spike] Investigate horizontal instance scaling

### ISSUE THREAD

**Issue by cweagans** [department-of-veterans-affairs/va.gov-cms#8734]
## Description
The CMS team completed #8443 as a quick-fix for what led to the outage on 3/11, but a better solution is to rely on autoscaling. We need to determine if there are any major changes required to support that. Considerations: availability zones (increased uptime), adjustments for downtime/after hours, cost savings, etc. 

## Acceptance Criteria 

- [x] Document and prioritize options to horizontally scale server load
	- see [here](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/8734#issuecomment-1119290107)
- [x] Create followup ticket(s) for implementation 
- [x] Evaluate https://github.com/department-of-veterans-affairs/va.gov-cms/issues/4459 and the child tickets and re-purpose or close as needed

### CMS Team
Please check the team(s) that will do this work.

- [x] `Platform CMS Team`
- [ ] `Sitewide crew `
  - [ ] `⭐️ Content ops`
  - [ ] `⭐️ CMS experience`
  - [ ] `⭐️ Public websites`
  - [ ] `⭐️ Facilities`
  - [ ] `⭐️ User support`

**Comment by cweagans** [comment]
I spent some time noodling on this this sprint + talked it through with @olivereri . What we landed on is a bit of a departure from our direction over the last six months or so, but I think it's the right thing to do.

To recap: back in November, we decided to stick with the BDR setup so that we could focus on accelerated publishing and other higher priority work. To enable that, we spent a couple of sprints upgrading to Amazon Linux 2 and upgrading PHP. That work was fairly straightforward and has served us well till now.

In looking into building horizontal scaling capabilities into the BRD system, it became clear that the amount of time that we'd have to invest into BRD far exceeds the amount of effort that we talked about before. Aside from the additional work required just to get the functionality, it will take additional time and tuning after we roll out the functionality. There were some pretty tangible benefits to sticking with BRD in November, but the gains are not as clear now, so it's time to start looking at divesting from BRD.

In parallel to this conversation, Eric has been working on bringing up Sorry Cypress on ECS + Fargate and the experience has been very positive. Within a week, he was able to bring up an application that he was not familiar with on brand new infrastructure. This is the kind of experience that we want for deploying the CMS as well, so our current thinking is to do the following (roughly in order):

1. Get Sorry Cypress fully up and running in a production-ready capacity on ECS and Fargate.
2. Containerize the CMS. This can happen without impacting BRD at all if we want to do the work that way. We can also consider doing the one-container-per-BRD-instance thing that BRD already supports if we want to run the container in production. The container images should be pushed to ECR.
3. Talk with people internally about potentially getting a VPC specific to the CMS with out own internet gateway, peering with the VA IP space, etc. This would allow us to set up a much simpler environment that doesn't rely on whatever the rest of the platform is doing (which is important because the CMS is an outlier in almost every way). If we can do this, then we should move our ECS + Fargate stuff to this new VPC.
4. Move CMS Test to ECS + Fargate using the new CMS container. Setting up the CMS to run on ECS and Fargate should include doing some basic load testing + configuring horizontal autoscaling.
5. Move CMS dev, staging, and prod to ECS and Fargate (in sequence)

Eric and I discussed the platform EKS offering as a potential path forward, however the problems that we identified in November are still present. Additionally, with the amount of turnover on the infrastructure team, we're not confident that moving to that hosting platform offers much benefit. Even if we aren't responsible for the care and feeding of that EKS cluster, if the CMS is down, it reflects poorly on us. Our end users see no distinction between a CMS outage and an EKS outage: if they can't edit content, it's the _CMS_ that's down. Given that we are fully responsible for the availability and performance of the CMS, we don't see it as a safe option to run the CMS on a platform that we are unable to effectively maintain (if for no other reason than political/ownership boundaries).

Moving toward a greatly simplified infrastructure helps us onboard new team members as well. As it stands right now, we have the choice of trying to get somebody up to speed with how BRD works (which is something that I admin that I'm still very shaky on, even after 8 months on the team) or getting somebody up to speed with how the VA has chosen to run Kubernetes (which is extremely convoluted compared to private sector setups that I've worked on). Neither of those options are particularly good. Getting somebody up to speed on a Fargate-deployed CMS would be much more straightforward.

**Comment by mchelen-gov** [comment]
interesting, recommend writing this up as an RFC. in particular some questions to explore & get feedback on:
- what are the platform and cms ato implications for each of these options?
- what is the technical loe for creation of vpc and va peering?
- what are the requirements/blockers for hosting on platform eks? if vets-api prod is hosted in eka with two nines uptime, is that sufficient?

**Comment by cweagans** [comment]
(note to self) related/interesting thread: https://dsva.slack.com/archives/CTYQL39FE/p1648136415886459

**Comment by olivereri** [comment]
Collecting information on ATO boundaries and IGWs
https://dsva.slack.com/archives/CJYRZK2HH/p1638897574340600

**Comment by cweagans** [comment]
Spike ticket for RFC: https://github.com/department-of-veterans-affairs/va.gov-cms/issues/9007
